src/cl.rs

```diff
@@ -350,33 +350,37 @@ impl<E: Engine> PublicKey<E> {
         message: &Vec<E::Fr>,
         signature: &Signature<E>,
     ) -> bool {
-        let mut L = E::G2::zero();
+        // zero is valid for any message, so never allow it
+        if signature.h == E::G1::zero() {
+            return false
```

Comment:
I know we don't optimize for constant-time, and I believe the underlying BLS12 implementation is not constant time, so this might be stupid. But shouldn't we avoid early returns? I know there are probably tons of places where we can do better on this, it's just a thought. (same comment for all the other places where you've added this check.